File: app/src/main/java/com/yhjoo/dochef/activities/NoticeActivity.java
package com.yhjoo.dochef.activities;

import android.os.Bundle;
import android.view.ViewGroup;

import androidx.recyclerview.widget.LinearLayoutManager;

import com.chad.library.adapter.base.entity.MultiItemEntity;
import com.yhjoo.dochef.App;
import com.yhjoo.dochef.R;
import com.yhjoo.dochef.adapter.NoticeListAdapter;
import com.yhjoo.dochef.databinding.ANoticeBinding;
import com.yhjoo.dochef.interfaces.RxRetrofitServices;
import com.yhjoo.dochef.model.ExpandContents;
import com.yhjoo.dochef.model.ExpandTitle;
import com.yhjoo.dochef.model.Notice;
import com.yhjoo.dochef.utils.DataGenerator;
import com.yhjoo.dochef.utils.RxRetrofitBuilder;

import java.util.ArrayList;

import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;

public class NoticeActivity extends BaseActivity {
    ANoticeBinding binding;

    RxRetrofitServices.BasicService basicService;
    NoticeListAdapter noticeListAdapter;

    ArrayList<MultiItemEntity> noticeList = new ArrayList<>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ANoticeBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        setSupportActionBar(binding.noticeToolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        basicService =
                RxRetrofitBuilder.create(this, RxRetrofitServices.BasicService.class);

        noticeListAdapter = new NoticeListAdapter(noticeList);
        binding.noticeRecycler.setLayoutManager(new LinearLayoutManager(this));
        binding.noticeRecycler.setAdapter(noticeListAdapter);
    }

    @Override
    protected void onResume() {
        super.onResume();

        if (App.isServerAlive()) {
            compositeDisposable.add(
                    basicService.getNotice()
                            .observeOn(AndroidSchedulers.mainThread())
                            .subscribe(response -> loadList(response.body()), RxRetrofitBuilder.defaultConsumer())
            );
        } else {
            ArrayList<Notice> response = DataGenerator.make(getResources(),
                    getResources().getInteger(R.integer.DATA_TYPE_NOTICE));
            loadList(response);
        }
    }

    void loadList(ArrayList<Notice> resList) {
        for (Notice item : resList) {
            ExpandTitle title = new ExpandTitle(item.title);
            title.addSubItem(new ExpandContents(item.contents, 0));
            noticeList.add(title);
        }

        noticeListAdapter.setNewData(noticeList);
        noticeListAdapter.setEmptyView(R.layout.rv_empty, (ViewGroup) binding.noticeRecycler.getParent());
    }
}

File: app/src/main/java/com/yhjoo/dochef/ui/activities/NoticeActivity.kt
package com.yhjoo.dochef.ui.activities

import android.os.Bundle
import android.util.Log
import androidx.recyclerview.widget.LinearLayoutManager
import com.chad.library.adapter.base.entity.MultiItemEntity
import com.yhjoo.dochef.App
import com.yhjoo.dochef.R
import com.yhjoo.dochef.data.DataGenerator
import com.yhjoo.dochef.data.model.ExpandContents
import com.yhjoo.dochef.data.model.ExpandTitle
import com.yhjoo.dochef.data.model.Notice
import com.yhjoo.dochef.databinding.ANoticeBinding
import com.yhjoo.dochef.ui.adapter.NoticeListAdapter
import com.yhjoo.dochef.utils.RetrofitBuilder
import com.yhjoo.dochef.utils.RetrofitServices.BasicService
import com.yhjoo.dochef.utils.Utils
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.util.*

class NoticeActivity : BaseActivity() {
    val binding: ANoticeBinding by lazy { ANoticeBinding.inflate(layoutInflater) }

    private lateinit var basicService: BasicService
    private lateinit var noticeListAdapter: NoticeListAdapter
    private var noticeList =  ArrayList<MultiItemEntity>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(binding.root)
        setSupportActionBar(binding.noticeToolbar)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)

        basicService = RetrofitBuilder.create(this, BasicService::class.java)

        noticeListAdapter = NoticeListAdapter(noticeList)
        binding.noticeRecycler.layoutManager = LinearLayoutManager(this)
        binding.noticeRecycler.adapter = noticeListAdapter
    }

    override fun onResume() {
        super.onResume()
        CoroutineScope(Dispatchers.Main).launch {
            runCatching {
                if (App.isServerAlive) {
                    val noticeResponse = basicService.getNotice()
                    loadList(noticeResponse.body()!!)
                } else {
                    val noticeResponse = withContext(Dispatchers.IO) {
                        DataGenerator.make<ArrayList<Notice>>(
                            resources,
                            resources.getInteger(R.integer.DATA_TYPE_NOTICE)
                        )
                    }
                    loadList(noticeResponse)
                }
            }
                .onSuccess { }
                .onFailure {
                    RetrofitBuilder.defaultErrorHandler(it)
                }
        }
    }

    private fun loadList(resList: ArrayList<Notice>) {
        for (item in resList) {
            val title = ExpandTitle(item.title)
            title.addSubItem(ExpandContents(item.contents, 0))
            noticeList.add(title)
        }
        noticeListAdapter.setNewData(noticeList as List<MultiItemEntity>)
    }
}